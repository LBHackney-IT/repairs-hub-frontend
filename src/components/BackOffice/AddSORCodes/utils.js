import { frontEndApiRequest } from '@/root/src/utils/frontEndApiClient/requests'

export const fetchContractors = () =>
  new Promise(async (resolve) => {
    const contractors = await frontEndApiRequest({
      method: 'get',
      path: '/api/contractors?getAllContractors=true',
    })

    resolve(contractors)
  })

export const fetchTrades = () =>
  new Promise(async (resolve) => {
    const trades = await frontEndApiRequest({
      method: 'get',
      path: '/api/schedule-of-rates/trades?getAllTrades=true',
    })

    resolve(trades)
  })

export const fetchContracts = (contractorReference) =>
  new Promise(async (resolve) => {
    const contracts = await frontEndApiRequest({
      method: 'get',
      path: `/api/backoffice/contracts?contractorReference=${contractorReference}`,
    })

    resolve(contracts)
  })

export const saveSorCodesToDatabase = async (data) => {
  await frontEndApiRequest({
    method: 'post',
    path: `/api/backoffice/sor-codes`,
    requestData: data,
  })
}

export const dataToRequestObject = (csvArray, contractReference, tradeCode) => {
  const sorCodes = csvArray.map((x) => ({
    code: x.Code,
    cost: parseFloat(x.Cost),
    standardMinuteValue: parseFloat(x.StandardMinuteValue),
    shortDescription: x.ShortDescription,
    longDescription: x.LongDescription,
  }))

  return {
    contractReference: contractReference,
    tradeCode: tradeCode,
    sorCodes: sorCodes,
    contract: 'placeholder',
  }
}
