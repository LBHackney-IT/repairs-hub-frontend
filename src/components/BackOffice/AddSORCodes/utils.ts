import { frontEndApiRequest } from '@/root/src/utils/frontEndApiClient/requests'

import { Contractor, Trade, NewSorCodesRequestObject } from './types'

export const fetchContractors = async () : Promise<Array<Contractor>>  => {
  return await frontEndApiRequest({
    method: 'get',
    path: '/api/contractors?getAllContractors=true',
    params: null,
    paramsSerializer: null,
    requestData: null
  })
}

export const fetchTrades = async () : Promise<Array<Trade>> => {
  return await frontEndApiRequest({
    method: 'get',
    path: '/api/schedule-of-rates/trades?getAllTrades=true',
    params: null,
    paramsSerializer: null,
    requestData: null
  })
}

export const fetchContracts = async (contractorReference: string) : Promise<Array<string>> => {
  return await frontEndApiRequest({
    method: 'get',
    path: `/api/backoffice/contracts?contractorReference=${contractorReference}`,
    params: null,
    paramsSerializer: null,
    requestData: null
  })
}

export const saveSorCodesToDatabase = async (data: NewSorCodesRequestObject): Promise<void> => {
  await frontEndApiRequest({
    method: 'post',
    path: `/api/backoffice/sor-codes`,
    requestData: data,
    params: null,
    paramsSerializer: null,
  })
}

export const dataToRequestObject = (csvArray: Array<any>, contractReference: string, tradeCode: string): NewSorCodesRequestObject => {
  const sorCodes = csvArray.map((x) => ({
    code: x.Code,
    cost: parseFloat(x.Cost),
    standardMinuteValue: parseFloat(x.StandardMinuteValue),
    shortDescription: x.ShortDescription,
    longDescription: x.LongDescription,
  }))

  return {
    contractReference: contractReference,
    tradeCode: tradeCode,
    sorCodes: sorCodes,
  }
}
