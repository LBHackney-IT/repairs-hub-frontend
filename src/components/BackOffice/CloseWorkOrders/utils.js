export const formatWorkOrderReferences = (workOrderReferences) => {
  return [
    ...new Set(
      workOrderReferences
        .split('\n')
        .map((x) => x.trim())
        .filter((x) => x)
    ),
  ]
}

const validateWorkOrderReference = (workOrderReference) => {
  const regex = /^\d{8}$/gm

  return regex.exec(workOrderReference) !== null
}

export const getInvalidWorkOrderReferences = (workOrderReferences) => {
  return workOrderReferences
    .filter((x) => !validateWorkOrderReference(x))
    .map((x) => `"${x}"`)
}

export const formatInvalidWorkOrderReferencesError = (
  invalidWorkOrderReferences
) => {
  return `Invalid WorkOrder Reference(s) entered: ${invalidWorkOrderReferences.join(
    ', '
  )}`
}

export const dateIsInFuture = (date) => {
  const today = new Date()

  return date > today
}
