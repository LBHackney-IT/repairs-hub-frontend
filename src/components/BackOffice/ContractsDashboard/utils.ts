import Contract from '@/root/src/models/contract'

export const today = new Date()

export function monthsOffset(months: number) {
  return new Date(
    today.getFullYear(),
    today.getMonth() + months,
    today.getDate()
  )
}

export const filterContractsByExpiryDate = (
  contracts: Contract[],
  months: number
) => {
  const contractExpiryCutOffDate = monthsOffset(months)

  const filteredContracts = contracts?.filter((contract) => {
    return (
      new Date(contract.terminationDate) > today &&
      new Date(contract.terminationDate) < contractExpiryCutOffDate
    )
  })
  return filteredContracts
}

export const filterRelativeInactiveContracts = (
  contracts: Contract[],
  yearAfter: string
) => {
  const relativeInactiveContracts = contracts?.filter(
    (contract) =>
      contract.terminationDate > yearAfter &&
      contract.terminationDate < new Date().toISOString()
  )
  return relativeInactiveContracts
}
