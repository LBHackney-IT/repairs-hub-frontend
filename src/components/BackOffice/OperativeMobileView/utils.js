export const filterOperatives = (operatives, operativeFilter) => {
  if (!operativeFilter) return operatives;

  return operatives.filter((x) =>
    x.id?.toString().includes(operativeFilter) ||
    x.name?.toLowerCase().includes(operativeFilter) ||
    x.payrollNumber?.toLowerCase().includes(operativeFilter)
  );
}
