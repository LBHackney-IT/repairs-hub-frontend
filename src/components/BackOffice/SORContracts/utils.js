export const sanitizeInput = value => {
    return value.trim().replaceAll(' ', '')
}

export const saveContractChangesToDatabase = async (data) => {
    await frontEndApiRequest({
        method: 'POST',
        path: `/api/backOffice/sor-contracts`,
        requestData: data,
      })
  }