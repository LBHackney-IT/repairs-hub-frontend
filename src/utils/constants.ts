export const PURDY_CONTRACTOR_REFERENCE = 'PUR'
export const MULTITRADE_TRADE_CODE = 'MU'
