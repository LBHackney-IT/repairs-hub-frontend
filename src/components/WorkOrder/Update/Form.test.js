import { render } from '@testing-library/react'
import WorkOrderUpdateForm from './Form'
import { PURDY_CONTRACTOR_REFERENCE } from '@/utils/constants'

jest.mock('next/router', () => ({
  useRouter() {
    return {
      pathname: '/work-orders/[id]/update',
    }
  },
}))

describe('WorkOrderUpdateForm component', () => {
  const props = {
    propertyReference: '000012345',
    originalTasks: [
      {
        code: 'DES5R006',
        quantity: 1,
        original: true,
        originalQuantity: 1,
      },
    ],
    latestTasks: [
      {
        code: 'DES5R006',
        quantity: 2,
        original: true,
        originalQuantity: 1,
      },
      {
        code: 'DES5R005',
        quantity: 1,
        original: false,
        originalQuantity: null,
      },
    ],
    addedTasks: [
      {
        code: 'DES5R004',
        quantity: 2,
        description: 'Immediate call outs',
      },
    ],
    onGetToSummary: jest.fn(),
    setVariationReason: jest.fn(),
    sorSearchRequest: jest.fn(),
    setSorCodeArrays: jest.fn(),
    variationReason: 'More work is necessary',
  }

  it('should render properly with unlimited characters for variation reason, as contractor is Purdy', () => {
    const { asFragment } = render(
      <WorkOrderUpdateForm
        propertyReference={props.propertyReference}
        originalTasks={props.originalTasks}
        latestTasks={props.latestTasks}
        addedTasks={props.addedTasks}
        onGetToSummary={props.onGetToSummary}
        setVariationReason={props.setVariationReason}
        variationReason={props.variationReason}
        contractor={{
          contractorReference: 'PCL',
          contractorName: 'Purdy',
          useExternalScheduleManager: false,
          canAssignOperative: false,
          perTradeAvailability: false,
          multiTradeEnabled: true,
        }}
        sorCodeArrays={[
          [
            {
              code: 'DES5R003',
              shortDescription: 'Immediate call outs',
              priority: {
                priorityCode: 1,
                description: '1 [I] IMMEDIATE',
              },
              cost: 0,
            },
            {
              code: 'DES5R004',
              shortDescription: 'Emergency call out',
              priority: {
                priorityCode: 2,
                description: '2 [E] EMERGENCY',
              },
            },
          ],
        ]}
        setSorCodeArrays={props.setSorCodeArrays}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render properly with 250 characters limit for variation reason', () => {
    const { asFragment } = render(
      <WorkOrderUpdateForm
        propertyReference={props.propertyReference}
        originalTasks={props.originalTasks}
        latestTasks={props.latestTasks}
        addedTasks={props.addedTasks}
        onGetToSummary={props.onGetToSummary}
        setVariationReason={props.setVariationReason}
        variationReason={props.variationReason}
        contractorReference="ABC"
        sorCodeArrays={[
          [
            {
              code: 'DES5R003',
              shortDescription: 'Immediate call outs',
              priority: {
                priorityCode: 1,
                description: '1 [I] IMMEDIATE',
              },
              cost: 0,
            },
            {
              code: 'DES5R004',
              shortDescription: 'Emergency call out',
              priority: {
                priorityCode: 2,
                description: '2 [E] EMERGENCY',
              },
            },
          ],
        ]}
        sorSearchRequest={props.sorSearchRequest}
        setSorCodeArrays={props.setSorCodeArrays}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render properly when there are no added tasks', () => {
    const { asFragment } = render(
      <WorkOrderUpdateForm
        propertyReference={props.propertyReference}
        originalTasks={props.originalTasks}
        latestTasks={props.latestTasks}
        addedTasks={[]}
        onGetToSummary={props.onGetToSummary}
        setVariationReason={props.setVariationReason}
        variationReason={props.variationReason}
        sorCodeArrays={[
          [
            {
              code: 'DES5R003',
              shortDescription: 'Immediate call outs',
              priority: {
                priorityCode: 1,
                description: '1 [I] IMMEDIATE',
              },
              cost: 0,
            },
            {
              code: 'DES5R004',
              shortDescription: 'Emergency call out',
              priority: {
                priorityCode: 2,
                description: '2 [E] EMERGENCY',
              },
            },
          ],
        ]}
        setSorCodeArrays={props.setSorCodeArrays}
        contractorReference={'ABC'}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
