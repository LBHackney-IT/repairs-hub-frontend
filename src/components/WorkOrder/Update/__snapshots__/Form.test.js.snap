// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkOrderUpdateForm component should render properly when there are no added tasks 1`] = `
<DocumentFragment>
  <form
    id="repair-request-form"
    role="form"
  >
    <div>
      <section
        class="section"
        id="original-rate-schedule-items"
      >
        <h2
          class="lbh-heading-h2"
        >
          Original tasks and SORS raised with work order
        </h2>
        <div
          class="govuk-grid-row original-rate-schedule-items align-items-center"
        >
          <div>
            <div
              class="govuk-grid-column-two-thirds"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="originalRateScheduleItems[0][code]-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="originalRateScheduleItems[0][code]"
                >
                  SOR code 
                </label>
                <input
                  class="govuk-input lbh-input undefined"
                  data-testid="originalRateScheduleItems[0][code]"
                  disabled=""
                  id="originalRateScheduleItems[0][code]"
                  name="originalRateScheduleItems[0][code]"
                  type="text"
                  value="DES5R006"
                />
              </div>
            </div>
            <div
              class="govuk-grid-column-one-third"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="originalRateScheduleItems[0][quantity]-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="originalRateScheduleItems[0][quantity]"
                >
                  Quantity 
                </label>
                <input
                  class="govuk-input lbh-input undefined"
                  data-testid="originalRateScheduleItems[0][quantity]"
                  disabled=""
                  id="originalRateScheduleItems[0][quantity]"
                  name="originalRateScheduleItems[0][quantity]"
                  type="text"
                  value="1"
                />
              </div>
            </div>
          </div>
        </div>
        <hr
          class="govuk-section-break govuk-section-break--l govuk-section-break--visible"
        />
      </section>
    </div>
    <div>
      <section
        class="section"
        id="existing-rate-schedule-items"
      >
        <h2
          class="lbh-heading-h2"
        >
          Latest tasks and SORS against the work order
        </h2>
        <div
          class="govuk-grid-row rate-schedule-items align-items-center"
        >
          <div>
            <div
              class="govuk-grid-column-two-thirds"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="sor-code-0-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="sor-code-0"
                >
                  SOR code 
                </label>
                <input
                  class="govuk-input lbh-input sor-code-0 govuk-!-width-full"
                  data-testid="sor-code-0"
                  disabled=""
                  id="sor-code-0"
                  name="sor-code-0"
                  type="text"
                  value="DES5R006"
                />
              </div>
              <input
                id="original-rate-schedule-item-0"
                name="original-rate-schedule-item-0"
                type="hidden"
                value=""
              />
            </div>
            <div
              class="govuk-grid-column-one-third"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="quantity-0-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="quantity-0"
                >
                  Quantity 
                </label>
                <input
                  class="govuk-input lbh-input quantity-0 govuk-!-width-full"
                  data-testid="quantity-0"
                  id="quantity-0"
                  name="quantity-0"
                  type="text"
                  value="2"
                />
              </div>
            </div>
          </div>
          <div>
            <div
              class="govuk-grid-column-two-thirds"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="sor-code-1-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="sor-code-1"
                >
                   
                </label>
                <input
                  class="govuk-input lbh-input sor-code-1 govuk-!-width-full"
                  data-testid="sor-code-1"
                  disabled=""
                  id="sor-code-1"
                  name="sor-code-1"
                  type="text"
                  value="DES5R005"
                />
              </div>
              <input
                id="original-rate-schedule-item-1"
                name="original-rate-schedule-item-1"
                type="hidden"
                value=""
              />
            </div>
            <div
              class="govuk-grid-column-one-third"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="quantity-1-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="quantity-1"
                >
                   
                </label>
                <input
                  class="govuk-input lbh-input quantity-1 govuk-!-width-full"
                  data-testid="quantity-1"
                  id="quantity-1"
                  name="quantity-1"
                  type="text"
                  value="1"
                />
              </div>
            </div>
          </div>
        </div>
        <hr
          class="govuk-section-break govuk-section-break--l govuk-section-break--visible"
        />
      </section>
    </div>
    <div />
    <div>
      <a
        class="lbh-link"
        href="#"
      >
        + Add another SOR code
      </a>
    </div>
    <div>
      <a
        class="lbh-link"
        href="#"
      >
        + Add multiple SOR codes
      </a>
    </div>
    <div
      class="govuk-form-group lbh-form-group"
      id="variationReason-form-group"
    >
      <label
        class="govuk-label lbh-label"
        for="variationReason"
      >
        Variation reason 
        <span
          class="govuk-required"
        >
          *
        </span>
      </label>
      <textarea
        aria-describedby="variationReason-hint variationReason-error"
        class="govuk-textarea lbh-textarea"
        data-testid="variationReason"
        id="variationReason"
        name="variationReason"
        placeholder="Write a reason for the variation..."
        rows="4"
      >
        More work is necessary
      </textarea>
    </div>
    <span
      aria-live="polite"
      class="govuk-hint govuk-character-count__message"
    >
      You have 250 characters remaining.
    </span>
    <div
      class="govuk-form-group lbh-form-group"
    >
      <button
        class="govuk-button lbh-button"
        data-module="govuk-button"
        type="submit"
      >
        Next
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`WorkOrderUpdateForm component should render properly with 250 characters limit for variation reason 1`] = `
<DocumentFragment>
  <form
    id="repair-request-form"
    role="form"
  >
    <div>
      <section
        class="section"
        id="original-rate-schedule-items"
      >
        <h2
          class="lbh-heading-h2"
        >
          Original tasks and SORS raised with work order
        </h2>
        <div
          class="govuk-grid-row original-rate-schedule-items align-items-center"
        >
          <div>
            <div
              class="govuk-grid-column-two-thirds"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="originalRateScheduleItems[0][code]-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="originalRateScheduleItems[0][code]"
                >
                  SOR code 
                </label>
                <input
                  class="govuk-input lbh-input undefined"
                  data-testid="originalRateScheduleItems[0][code]"
                  disabled=""
                  id="originalRateScheduleItems[0][code]"
                  name="originalRateScheduleItems[0][code]"
                  type="text"
                  value="DES5R006"
                />
              </div>
            </div>
            <div
              class="govuk-grid-column-one-third"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="originalRateScheduleItems[0][quantity]-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="originalRateScheduleItems[0][quantity]"
                >
                  Quantity 
                </label>
                <input
                  class="govuk-input lbh-input undefined"
                  data-testid="originalRateScheduleItems[0][quantity]"
                  disabled=""
                  id="originalRateScheduleItems[0][quantity]"
                  name="originalRateScheduleItems[0][quantity]"
                  type="text"
                  value="1"
                />
              </div>
            </div>
          </div>
        </div>
        <hr
          class="govuk-section-break govuk-section-break--l govuk-section-break--visible"
        />
      </section>
    </div>
    <div>
      <section
        class="section"
        id="existing-rate-schedule-items"
      >
        <h2
          class="lbh-heading-h2"
        >
          Latest tasks and SORS against the work order
        </h2>
        <div
          class="govuk-grid-row rate-schedule-items align-items-center"
        >
          <div>
            <div
              class="govuk-grid-column-two-thirds"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="sor-code-0-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="sor-code-0"
                >
                  SOR code 
                </label>
                <input
                  class="govuk-input lbh-input sor-code-0 govuk-!-width-full"
                  data-testid="sor-code-0"
                  disabled=""
                  id="sor-code-0"
                  name="sor-code-0"
                  type="text"
                  value="DES5R006"
                />
              </div>
              <input
                id="original-rate-schedule-item-0"
                name="original-rate-schedule-item-0"
                type="hidden"
                value=""
              />
            </div>
            <div
              class="govuk-grid-column-one-third"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="quantity-0-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="quantity-0"
                >
                  Quantity 
                </label>
                <input
                  class="govuk-input lbh-input quantity-0 govuk-!-width-full"
                  data-testid="quantity-0"
                  id="quantity-0"
                  name="quantity-0"
                  type="text"
                  value="2"
                />
              </div>
            </div>
          </div>
          <div>
            <div
              class="govuk-grid-column-two-thirds"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="sor-code-1-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="sor-code-1"
                >
                   
                </label>
                <input
                  class="govuk-input lbh-input sor-code-1 govuk-!-width-full"
                  data-testid="sor-code-1"
                  disabled=""
                  id="sor-code-1"
                  name="sor-code-1"
                  type="text"
                  value="DES5R005"
                />
              </div>
              <input
                id="original-rate-schedule-item-1"
                name="original-rate-schedule-item-1"
                type="hidden"
                value=""
              />
            </div>
            <div
              class="govuk-grid-column-one-third"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="quantity-1-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="quantity-1"
                >
                   
                </label>
                <input
                  class="govuk-input lbh-input quantity-1 govuk-!-width-full"
                  data-testid="quantity-1"
                  id="quantity-1"
                  name="quantity-1"
                  type="text"
                  value="1"
                />
              </div>
            </div>
          </div>
        </div>
        <hr
          class="govuk-section-break govuk-section-break--l govuk-section-break--visible"
        />
      </section>
    </div>
    <div>
      <div
        class="rate-schedule-item govuk-!-margin-top-0"
      >
        <div
          class="govuk-form-group lbh-form-group rate-schedule-item--sor-code"
          id="rateScheduleItems[undefined][code]-form-group"
        >
          <label
            class="govuk-label lbh-label"
            for="rateScheduleItems[undefined][code]"
          >
            SOR Code 
            <span
              class="govuk-required"
            >
              *
            </span>
             
            <span>
               - Search by code or description
            </span>
          </label>
          <span
            class="govuk-hint lbh-hint"
            id="rateScheduleItems[undefined][code]-hint"
          >
            Enter three characters to view results
          </span>
          <input
            aria-describedby="rateScheduleItems[undefined][code]-hint"
            autocomplete="off"
            class="govuk-select lbh-select govuk-!-margin-top-0 govuk-!-width-full"
            data-testid="rateScheduleItems[undefined][code]"
            id="rateScheduleItems[undefined][code]"
            list="autocomplete-list-rateScheduleItems[undefined][code]"
            name="rateScheduleItems[undefined][code]"
            value="DES5R004 - Immediate call outs"
          />
          <datalist
            id="autocomplete-list-rateScheduleItems[undefined][code]"
          >
            <option
              value="DES5R003 - Immediate call outs"
            />
            <option
              value="DES5R004 - Emergency call out"
            />
          </datalist>
        </div>
        <input
          id="rateScheduleItems[undefined][description]"
          name="rateScheduleItems[undefined][description]"
          type="hidden"
          value="Immediate call outs"
        />
        <input
          id="rateScheduleItems[undefined][cost]"
          name="rateScheduleItems[undefined][cost]"
          type="hidden"
          value=""
        />
        <div
          class="govuk-form-group lbh-form-group rate-schedule-item--quantity"
          id="rateScheduleItems[undefined][quantity]-form-group"
        >
          <label
            class="govuk-label lbh-label"
            for="rateScheduleItems[undefined][quantity]"
          >
            Quantity 
            <span
              class="govuk-required"
            >
              *
            </span>
          </label>
          <input
            class="govuk-input lbh-input govuk-!-width-full"
            data-testid="rateScheduleItems[undefined][quantity]"
            id="rateScheduleItems[undefined][quantity]"
            name="rateScheduleItems[undefined][quantity]"
            type="text"
            value="2"
          />
        </div>
        <div
          class="remove-rate-schedule-item govuk-!-margin-0"
        >
          <button
            class="lbh-link"
            id="remove-rate-schedule-item-undefined"
            type="button"
          >
            Remove
          </button>
        </div>
      </div>
    </div>
    <div>
      <a
        class="lbh-link"
        href="#"
      >
        + Add another SOR code
      </a>
    </div>
    <div>
      <a
        class="lbh-link"
        href="#"
      >
        + Add multiple SOR codes
      </a>
    </div>
    <div
      class="govuk-form-group lbh-form-group"
      id="variationReason-form-group"
    >
      <label
        class="govuk-label lbh-label"
        for="variationReason"
      >
        Variation reason 
        <span
          class="govuk-required"
        >
          *
        </span>
      </label>
      <textarea
        aria-describedby="variationReason-hint variationReason-error"
        class="govuk-textarea lbh-textarea"
        data-testid="variationReason"
        id="variationReason"
        name="variationReason"
        placeholder="Write a reason for the variation..."
        rows="4"
      >
        More work is necessary
      </textarea>
    </div>
    <span
      aria-live="polite"
      class="govuk-hint govuk-character-count__message"
    >
      You have 250 characters remaining.
    </span>
    <div
      class="govuk-form-group lbh-form-group"
    >
      <button
        class="govuk-button lbh-button"
        data-module="govuk-button"
        type="submit"
      >
        Next
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`WorkOrderUpdateForm component should render properly with unlimited characters for variation reason, as contractor is Purdy 1`] = `
<DocumentFragment>
  <form
    id="repair-request-form"
    role="form"
  >
    <div>
      <section
        class="section"
        id="original-rate-schedule-items"
      >
        <h2
          class="lbh-heading-h2"
        >
          Original tasks and SORS raised with work order
        </h2>
        <div
          class="govuk-grid-row original-rate-schedule-items align-items-center"
        >
          <div>
            <div
              class="govuk-grid-column-two-thirds"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="originalRateScheduleItems[0][code]-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="originalRateScheduleItems[0][code]"
                >
                  SOR code 
                </label>
                <input
                  class="govuk-input lbh-input undefined"
                  data-testid="originalRateScheduleItems[0][code]"
                  disabled=""
                  id="originalRateScheduleItems[0][code]"
                  name="originalRateScheduleItems[0][code]"
                  type="text"
                  value="DES5R006"
                />
              </div>
            </div>
            <div
              class="govuk-grid-column-one-third"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="originalRateScheduleItems[0][quantity]-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="originalRateScheduleItems[0][quantity]"
                >
                  Quantity 
                </label>
                <input
                  class="govuk-input lbh-input undefined"
                  data-testid="originalRateScheduleItems[0][quantity]"
                  disabled=""
                  id="originalRateScheduleItems[0][quantity]"
                  name="originalRateScheduleItems[0][quantity]"
                  type="text"
                  value="1"
                />
              </div>
            </div>
          </div>
        </div>
        <hr
          class="govuk-section-break govuk-section-break--l govuk-section-break--visible"
        />
      </section>
    </div>
    <div>
      <section
        class="section"
        id="existing-rate-schedule-items"
      >
        <h2
          class="lbh-heading-h2"
        >
          Latest tasks and SORS against the work order
        </h2>
        <div
          class="govuk-grid-row rate-schedule-items align-items-center"
        >
          <div>
            <div
              class="govuk-grid-column-two-thirds"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="sor-code-0-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="sor-code-0"
                >
                  SOR code 
                </label>
                <input
                  class="govuk-input lbh-input sor-code-0 govuk-!-width-full"
                  data-testid="sor-code-0"
                  disabled=""
                  id="sor-code-0"
                  name="sor-code-0"
                  type="text"
                  value="DES5R006"
                />
              </div>
              <input
                id="original-rate-schedule-item-0"
                name="original-rate-schedule-item-0"
                type="hidden"
                value=""
              />
            </div>
            <div
              class="govuk-grid-column-one-third"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="quantity-0-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="quantity-0"
                >
                  Quantity 
                </label>
                <input
                  class="govuk-input lbh-input quantity-0 govuk-!-width-full"
                  data-testid="quantity-0"
                  id="quantity-0"
                  name="quantity-0"
                  type="text"
                  value="2"
                />
              </div>
            </div>
          </div>
          <div>
            <div
              class="govuk-grid-column-two-thirds"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="sor-code-1-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="sor-code-1"
                >
                   
                </label>
                <input
                  class="govuk-input lbh-input sor-code-1 govuk-!-width-full"
                  data-testid="sor-code-1"
                  disabled=""
                  id="sor-code-1"
                  name="sor-code-1"
                  type="text"
                  value="DES5R005"
                />
              </div>
              <input
                id="original-rate-schedule-item-1"
                name="original-rate-schedule-item-1"
                type="hidden"
                value=""
              />
            </div>
            <div
              class="govuk-grid-column-one-third"
            >
              <div
                class="govuk-form-group lbh-form-group"
                id="quantity-1-form-group"
              >
                <label
                  class="govuk-label lbh-label"
                  for="quantity-1"
                >
                   
                </label>
                <input
                  class="govuk-input lbh-input quantity-1 govuk-!-width-full"
                  data-testid="quantity-1"
                  id="quantity-1"
                  name="quantity-1"
                  type="text"
                  value="1"
                />
              </div>
            </div>
          </div>
        </div>
        <hr
          class="govuk-section-break govuk-section-break--l govuk-section-break--visible"
        />
      </section>
    </div>
    <div>
      <div
        class="rate-schedule-item govuk-!-margin-top-0"
      >
        <div
          class="govuk-form-group lbh-form-group rate-schedule-item--sor-code"
          id="rateScheduleItems[undefined][code]-form-group"
        >
          <label
            class="govuk-label lbh-label"
            for="rateScheduleItems[undefined][code]"
          >
            SOR Code 
            <span
              class="govuk-required"
            >
              *
            </span>
             
            <span>
               - Search by code or description
            </span>
          </label>
          <input
            autocomplete="off"
            class="govuk-select lbh-select govuk-!-margin-top-0 govuk-!-width-full"
            data-testid="rateScheduleItems[undefined][code]"
            id="rateScheduleItems[undefined][code]"
            list="autocomplete-list-rateScheduleItems[undefined][code]"
            name="rateScheduleItems[undefined][code]"
            value="DES5R004 - Immediate call outs"
          />
          <datalist
            id="autocomplete-list-rateScheduleItems[undefined][code]"
          >
            <option
              value="DES5R003 - Immediate call outs"
            />
            <option
              value="DES5R004 - Emergency call out"
            />
          </datalist>
        </div>
        <input
          id="rateScheduleItems[undefined][description]"
          name="rateScheduleItems[undefined][description]"
          type="hidden"
          value="Immediate call outs"
        />
        <input
          id="rateScheduleItems[undefined][cost]"
          name="rateScheduleItems[undefined][cost]"
          type="hidden"
          value=""
        />
        <div
          class="govuk-form-group lbh-form-group rate-schedule-item--quantity"
          id="rateScheduleItems[undefined][quantity]-form-group"
        >
          <label
            class="govuk-label lbh-label"
            for="rateScheduleItems[undefined][quantity]"
          >
            Quantity 
            <span
              class="govuk-required"
            >
              *
            </span>
          </label>
          <input
            class="govuk-input lbh-input govuk-!-width-full"
            data-testid="rateScheduleItems[undefined][quantity]"
            id="rateScheduleItems[undefined][quantity]"
            name="rateScheduleItems[undefined][quantity]"
            type="text"
            value="2"
          />
        </div>
        <div
          class="remove-rate-schedule-item govuk-!-margin-0"
        >
          <button
            class="lbh-link"
            id="remove-rate-schedule-item-undefined"
            type="button"
          >
            Remove
          </button>
        </div>
      </div>
    </div>
    <div>
      <a
        class="lbh-link"
        href="#"
      >
        + Add another SOR code
      </a>
    </div>
    <div>
      <a
        class="lbh-link"
        href="#"
      >
        + Add multiple SOR codes
      </a>
    </div>
    <div
      class="govuk-form-group lbh-form-group"
      id="variationReason-form-group"
    >
      <label
        class="govuk-label lbh-label"
        for="variationReason"
      >
        Variation reason 
        <span
          class="govuk-required"
        >
          *
        </span>
      </label>
      <textarea
        aria-describedby="variationReason-hint variationReason-error"
        class="govuk-textarea lbh-textarea"
        data-testid="variationReason"
        id="variationReason"
        name="variationReason"
        placeholder="Write a reason for the variation..."
        rows="4"
      >
        More work is necessary
      </textarea>
    </div>
    <span
      aria-live="polite"
      class="govuk-hint govuk-character-count__message"
    >
      You have 250 characters remaining.
    </span>
    <div
      class="govuk-form-group lbh-form-group"
    >
      <button
        class="govuk-button lbh-button"
        data-module="govuk-button"
        type="submit"
      >
        Next
      </button>
    </div>
  </form>
</DocumentFragment>
`;
